lib/widgets/set_status.dart

```diff
@@ -0,0 +1,339 @@
+import 'package:flutter/material.dart';
+
+import '../api/model/model.dart';
+import '../api/route/users.dart';
+import '../basic.dart';
+import '../generated/l10n/zulip_localizations.dart';
+import '../log.dart';
+import 'app_bar.dart';
+import 'color.dart';
+import 'content.dart';
+import 'emoji_reaction.dart';
+import 'icons.dart';
+import 'inset_shadow.dart';
+import 'page.dart';
+import 'store.dart';
+import 'text.dart';
+import 'theme.dart';
+
+class SetStatusPage extends StatefulWidget {
+  const SetStatusPage({super.key, required this.oldStatus});
+
+  final UserStatus oldStatus;
+
+  static AccountRoute<void> buildRoute({
+    required BuildContext context,
+    required UserStatus oldStatus,
+  }) {
+    return MaterialAccountWidgetRoute(context: context,
+      page: SetStatusPage(oldStatus: oldStatus));
+  }
+
+  @override
+  State<SetStatusPage> createState() => _SetStatusPageState();
+}
+
+class _SetStatusPageState extends State<SetStatusPage> {
+  late final TextEditingController statusTextController;
+  late final ValueNotifier<UserStatusChange> statusChange;
+
+  UserStatus get oldStatus => widget.oldStatus;
+  UserStatus get newStatus => statusChange.value.apply(widget.oldStatus);
+
+  @override
+  void initState() {
+    super.initState();
+    statusTextController = TextEditingController(text: oldStatus.text)
+      ..addListener(() {
+        final trimmedValue = statusTextController.text.trim();
+        final text = trimmedValue.isNotEmpty ? trimmedValue : null;
+
+        // Ignore updating [statusChange] for the additional updates with the
+        // same value from TextField. For example, when a character is deleted.
+        if (text == newStatus.text) return;
+
+        statusChange.value = statusChange.value.copyWith(
+          text: toOption(recent: text, old: oldStatus.text),
+        );
+      });
+    statusChange =
+      ValueNotifier(UserStatusChange(text: OptionNone(), emoji: OptionNone()))
+        ..addListener(() {
+          final text = statusChange.value.text.or(oldStatus.text) ?? '';
+
+          // Ignore updating the status text field if it already has the same
+          // text. It can happen in the following cases:
+          //   1. Only the emoji is changed.
+          //   2. The same status is chosen consecutively from the suggested
+          //      statuses list.
+          //   3. This listener is called as a result of the change in status
+          //      text field.
+          if (text == statusTextController.text) return;
+
+          statusTextController.text = text;
+        });
+  }
+
+  @override
+  void dispose() {
+    statusTextController.dispose();
+    statusChange.dispose();
+    super.dispose();
+  }
+
+  List<UserStatus> statusSuggestions(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+    final localizations = ZulipLocalizations.of(context);
+
+    final values = [
+      (localizations.userStatusBusy,            '1f6e0', 'working_on_it'),
```

Comment:
If the data we have from the server is missing one of these emoji, I think better to leave it out from the suggestions than to try to fall back to a hard-coded name. It's likely in that case that the hard-coded name won't work either.

(continuing from https://github.com/zulip/zulip-flutter/pull/1701#discussion_r2211785018)